### ISSUE TITLE
github Failed to log in: invalid state parameter 

### ISSUE THREAD

**Issue by cg-cnu** [rust-lang/crates.io#1142]
As mentioned in #1077 

on registering and login in its throwing an error...
`Failed to log in: invalid state parameter`
Tried revoking the access and loging in from incognito. No success.
Let me know if I can help in debugging it. 
Btw, am on on mac sierra with chrome.

**Comment by carols10cents** [comment]
Hi @cg-cnu, sorry you're running into this and sorry I haven't gotten back to you!

Is this still happening? Could you try clearing your cookies for crates.io and see if that changes anything?

How much debugging are you willing to do? :) Could you try setting up a local instance of crates.io and a GitHub oAuth application and see if the same problem happens with you when signing into your locally running instance? There are instructions for setting up a development environment here: https://github.com/rust-lang/crates.io/blob/master/docs/CONTRIBUTING.md#setting-up-a-development-environment and instructions on setting up GitHub integration in this section: https://github.com/rust-lang/crates.io/blob/master/docs/CONTRIBUTING.md#publishing-a-crate-to-your-local-cratesio

Some background/guesses/things I'm thinking about:

I'm not really sure if this is a crates.io problem, a github problem, a network problem, something else-- this error message comes from [this spot](https://github.com/rust-lang/crates.io/blob/601ed3ca93bd92c66d5b4194b4bb0c01c878a5cb/src/user/mod.rs#L315) because the state parameter (generated [here](https://github.com/rust-lang/crates.io/blob/601ed3ca93bd92c66d5b4194b4bb0c01c878a5cb/src/user/mod.rs#L258) and sent to github) returned from github is not matching the state parameter we saved in the current session. Some things I'd like to know if you're able to reproduce this with a locally running instance:

- Are both `Some(&state[..])` and `session_state` non-empty strings?
- Are there any special characters in either of the state values that are being encoded incorrectly?
- Or are they both random looking, very different strings? If this is the case, I'm not sure where to go from there :-/

**Comment by cg-cnu** [comment]
Hey @carols10cents 
Thanks for the detailed explanation. I would setup the environment and do the debugging over the weekend. I will get back to you if I have any questions. Thanks 😄

**Comment by tuwid** [comment]
@carols10cents having the same issue with Azure AD any updates ?

**Comment by ChristopherMacGown** [comment]
I can also reproduce this on the public site. However, I wasn't able to reproduce it locally against the `rust-lang/crates.io` `master` branch with either Safari or Chrome (11.0.2 and 63.0.3239.132 respectively). The [logs](https://gist.github.com/ChristopherMacGown/a4cca486552c6ed0b40c842a511aba09) suggest that the parameters GitHub is returning are correct.

**Comment by nghenglim** [comment]
Same here, unable to login to crates.io. In the chrome dev tools:
```
Request URL:https://github.com/login/oauth/authorize?client_id=MY_CLIENT_ID&state=MY_STATE&scope=read%3Aorg
Request Method:GET
Status Code:302 Found
Remote Address:192.30.255.112:443
Referrer Policy:no-referrer-when-downgrade
```
```
Request URL: https://crates.io/authorize/github?code=MY_CODE&state=MY_STATE
Request Method:GET
Status Code:200 OK
Remote Address:23.23.136.239:443
Referrer Policy:no-referrer-when-downgrade
```
```
Request URL:https://crates.io/authorize?code=MY_CODE&state=MY_STATE
Request Method:GET
Status Code:404 Not Found
Remote Address:23.23.136.239:443
Referrer Policy:no-referrer-when-downgrade
```
Base on the result above, I will assume that something wrong inside the crates.io database there. Just guessing, probably some table failed to update during attempt to signing me up.

**Comment by nghenglim** [comment]
Okay found out the problem! Looking at the code I figured out only name / email field which crates.io get from github could be wrong.

In github I actually set my public email to NA (because I don't want people to see my actual email there. To check out whether is email caused the problem I set it to my actual verified email and try login! then success!

However I don't think crates.io should be using our public github email as a reference field (and somemore tries to send verification email to it!)

tl;dr: Just change github email to true email then it fixed the problem.

**Comment by genbattle** [comment]
Adding my +1 to this issue. Strangely I seem to have my email publicly displayed, so I'm not sure exactly why I'm getting this error when trying to log into crates.io. I've never logged into crates.io before.

**Comment by genbattle** [comment]
Editing my profile in Github (even though I don't think I changed anything) seems to have solved the issue for me? Possibly there's some sort of grandfathering or backend migration that kicked in when I re-saved my profile.

**Comment by prewk** [comment]
I tried editing the profile as suggested by @genbattle but to no avail. Still getting the error.

**Comment by prewk** [comment]
Suddenly works now!

**Comment by oli-obk** [comment]
I just ran into the same problem, and it ended up being Ghostery (tracking blocking) which broke the login. Disabling Ghostery for crates.io solves the problem

**Comment by cbiffle** [comment]
This is still very much an issue -- crates.io fails with the "invalid state parameter" message for me unless I put a public email address up. If this is a requirement, y'all should at least improve the error message.

**Comment by carols10cents** [comment]
@cbiffle it is not a requirement. We've actually [merged a fix for this](https://github.com/rust-lang/crates.io/pull/1622), we just need to deploy it. Thank you for your patience!

**Comment by AurevoirXavier** [comment]
Still got this in 2020.

**Comment by kingdream123** [comment]
Still got this in 2024.
